internal/ingress/annotations/class/main.go

```diff
@@ -37,30 +38,30 @@ var (
 	IngressClass = "nginx"
 )
 
-// IsValid returns true if the given Ingress either doesn't specify
-// the ingress.class annotation, or it's set to the configured in the
-// ingress controller.
+// IsValid returns true if the given Ingress specify the ingress.class
+// annotation or IngressClassName resource for Kubernetes >= v1.18
 func IsValid(ing *networking.Ingress) bool {
-	className := ing.Spec.IngressClassName
+	// 1. with annotation
+	ingress, ok := ing.GetAnnotations()[IngressKey]
+	if ok {
+		// empty annotation and same annotation on ingress
+		if ingress == "" && IngressClass == DefaultClass {
```

Comment:
My interpretation of https://github.com/kubernetes/kubernetes/blob/d911254debefb5b630498aa3552899ee2f99f22e/pkg/apis/networking/types.go#L276 is that the controller should accept ingresses without set annotation only if configured `IngressClass` for the controller has `ingressclass.kubernetes.io/is-default-class` set to `true`.

I also assume the API won't allow more than one `IngressClass` resource to have that annotation set to `true` in a given cluster, which together with validation of unique `IngressClass` name will ensure there's no two controller in the cluster fighting over the same Ingress resource.

Reply:
We don't need to test that. The annotation is handled by an admission controller in k8s itself.

``` 
echo "
apiVersion: networking.k8s.io/v1beta1
kind: IngressClass
metadata:
  name: nginx
  annotations:
    ingressclass.kubernetes.io/is-default-class: 'true'
spec:
  controller: k8s.io/ingress-nginx
" | kubectl apply -f

echo "
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: ingress
spec:
  rules:
  - host: foo.bar.com
    http:
      paths:
      - path: /
        backend:
          serviceName: http-svc
          servicePort: 80
" | kubectl apply -f -
```

IngressClass with the annotation adds the `ingressClassName` in the ingress with value `nginx`.

Reply:
> When a
// single IngressClass resource has this annotation set to true, new Ingress
// resources without a class specified will be assigned this default class.

The behavior is to assign the IngressClass to new ingress resources if the class contains the annotation.